GiniVision/Classes/ToolTipView.swift

```diff
@@ -0,0 +1,300 @@
+    fileprivate var tipContainer: UIView
+    
+    var beforeDismiss: (() -> ())?
+    
+    init(text:String, textColor:UIColor, font:UIFont, backgroundColor: UIColor, closeButtonColor: UIColor, referenceView: UIView, superView:UIView, position: ToolTipPosition) {
+        
+        self.text = text
+        self.referenceView = referenceView
+        self.toolTipPosition = position
+        self.textLabel = UILabel()
+        self.closeButton = UIButton()
+        self.tipContainer = UIView()
+        self.arrowView = ToolTipView.arrow(withHeight: arrowHeight, width: arrowWidth, color: .white, position: position)
+        
+        super.init(frame: .zero)
+        superView.addSubview(self)
+        alpha = 0
+        
+        self.textSize = size(forText: text, withFont: font)
+        self.addTipContainer(backgroundColor: backgroundColor)
+        self.addTextLabel(withText: text, textColor: textColor, font: font)
+        self.addCloseButton(withColor:closeButtonColor)
+        self.addArrow()
+        self.addShadow()
+        
+        self.arrangeFrame(withSuperView: superView)
+        self.arrangeArrow(withSuperView: superView)
+        self.setupConstraints()
+    }
+    
+    override func layoutSubviews() {
+        super.layoutSubviews()
+        self.arrangeArrow(withSuperView: superview)
+    }
+    
+    func arrangeViews() {
+        self.arrangeFrame(withSuperView: superview)
+        self.arrangeArrow(withSuperView: superview)
+    }
+    
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("NSCoding not supported. Use init(text) instead!")
+    }
+    
+    // MARK: Add views
+    fileprivate func addTipContainer(backgroundColor color:UIColor) {
+        self.addSubview(tipContainer)
+        self.tipContainer.backgroundColor = color
+    }
+    
+    fileprivate func addTextLabel(withText text:String, textColor:UIColor, font:UIFont) {
+        textLabel.text = text
+        textLabel.textColor = textColor
+        textLabel.font = font
+        textLabel.numberOfLines = 0
+        tipContainer.addSubview(textLabel)
+    }
+    
+    fileprivate func addArrow() {
+        arrowView.frame.origin = CGPoint(x: 0, y: self.frame.height)
+        self.addSubview(arrowView)
+    }
+    
+    fileprivate func addCloseButton(withColor color:UIColor) {
+        let image = UIImageNamedPreferred(named: "toolTipCloseButton")
+        closeButton.setImage(image, for: .normal)
+        closeButton.tintColor = color
+        closeButton.addTarget(self, action: #selector(closeAction), for: .touchUpInside)
+        tipContainer.addSubview(closeButton)
+    }
+    
+    fileprivate func addShadow() {
+        self.layer.shadowOffset = CGSize(width: 0, height: 2)
+        self.layer.shadowRadius = 0.8
+        self.layer.shadowOpacity = 0.2
+        self.layer.shadowColor = UIColor.black.cgColor
+    }
+    
+    // MARK: Arrange views
+    fileprivate func arrangeFrame(withSuperView superView:UIView?) {
```

Comment:
This method is a monster... and difficult to read. Lets discuss it.

Reply:
I found a bug with the tip position (an edge case) that makes me simplify this method, so maybe now it is easier to read.